interfaces/prompting/patterns.go

```diff
@@ -0,0 +1,116 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"fmt"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+// Limit the number of groups which are allowed to occur in a pattern.
+// This number includes all groups, including those nested or in series.
+const maxGroupsInPattern = 10
+
+// ValidatePathPattern returns nil if the pattern is valid, otherwise an error.
+func ValidatePathPattern(pattern string) error {
+	if !doublestar.ValidatePattern(pattern) {
+		return fmt.Errorf("invalid path pattern: %q", pattern)
+	}
+	if pattern == "" || pattern[0] != '/' {
+		return fmt.Errorf("invalid path pattern: must start with '/': %q", pattern)
+	}
+	depth := 0
+	totalGroups := 0
+	reader := strings.NewReader(pattern)
+	for {
+		r, _, err := reader.ReadRune()
+		if err != nil {
+			// No more runes
+			break
+		}
+		switch r {
+		case '{':
```

Comment:
I think I'm confused even more now. The spec does not mention `{}` for path patterns, so why validate this? Also, there's inconsistency between ValidatePathPattern and MatchPathPatter, where the latter is oblivious to `{}`, which means that a definition of a valid pattern is different for each of those calls.

Reply:
`MatchPathPattern` directly calls `doublestar.Match` to do matching (discussed some here: https://docs.google.com/document/d/1tBnefdukP69EUJOlH8bgD2hrvZCYoE8-1ZlqRRYlOqc/edit?pli=1#heading=h.a0e0vfqho1ip). This should be called on expanded patterns, but this isn't technically necessary, since `doublestar.Match` happily matches patterns with `'{'` `'}'` groups in them (see: https://pkg.go.dev/github.com/bmatcuk/doublestar#Match). Path pattern validation should occur on patterns *before* they are expanded, so they may contain `'{'` `'}'`-defined groups.